content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,243 @@
+
+Secondly, when **the code is short, the bugs are nowhere to hide.** Parameter causes bugs easily because it needs extra concern. Reducing parameters can make the method more clear, and show the intention explicitly. It can help us to maintain the code easily. When we try to reduce the parameter, we are actually trying to reduce the potential bugs at the same time.
+
+Now we know it's better to reduce the parameter, but how could we do it? Let's check the parameter again.
+
+## Data clump
+When we find out the parameters are **a group of data that always come together**, and it would make no sense when missing one of them, we can call it a set of **Data Clump**. For example, a pair of start date and end date, or in this example, first name and last name.
+
+In the OOP world, the **object is the container of data, and class is the container of method** (behavior of object). It's cheap to create many small objects to manage and operate the serial data. Additionally, we can define the method on the object and reuse it when necessary.
+
+Let's try to create a new object to contain first name and last name.
+
+```rb
+class Author
+  attr_reader :first_name, :last_name
+ 
+  def initialize(first_name, last_name)
+    @first_name = first_name
+    @last_name = last_name
+  end
+ 
+  def full_name
+    "#{last_name}, #{first_name}"
+  end
+end
+```
+The Author class also has two attributes, and the author object has one instance method call `full_name`. Now we can use the author object to replace the first name and last name.
+
+#### Before
+```rb
+def summary(first_name, last_name)
+  "#{title} - #{last_name}, #{first_name} (#{publish_year})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary("Bater", "Chen")
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+#### After
+```rb
+def summary(author)
+  "#{title} - #{author.full_name} (#{publish_year})"
+end
+ 
+book = Book.new("Ruby is the best", "2020")
+book.summary(Author.new("Bater", "Chen"))
+=> "Ruby is the best - Chen, Bater (2020)"
+```
+The output of `summary` is the same, but the parameter reduces from two strings to one object. Well, you may doubt that it doesn't really matter, but imagine that you can follow the same pattern to reduce the parameter from 10 data into 1 object. That would be a huge impact.
+
+## Law of Demeter
+You may also notice that I relocate the logic of full name from book class to author class because a full name can perform independently without interaction with a book object. An **Object should know less about each other**, it is known as the [Law of Demeter][2] or "Least Knowledge Principle".
```

Comment:
```suggestion
You may also have noticed that I relocated the logic of full name from book class to author class because a full name can perform independently without interaction with the book object. An **object should know less about each other** is said by the [Law of Demeter][2] or the "Least Knowledge Principle".
```